merlion/models/layers.py

```diff
@@ -0,0 +1,253 @@
+        # Dynamically inherit from the appropriate kind of base model
+        original_cls = cls
+        config = cls._resolve_args(config=config, model=model, **kwargs)
+        if isinstance(config.model, ForecastingDetectorBase):
+            cls = cls.__class__(cls.__name__, (cls, LayeredForecastingDetector), {})
+            setattr(cls, "_original_cls", original_cls)
+        elif isinstance(config.model, ForecasterBase):
+            cls = cls.__class__(cls.__name__, (cls, LayeredForecaster), {})
+            setattr(cls, "_original_cls", original_cls)
+        elif isinstance(config.model, DetectorBase):
+            cls = cls.__class__(cls.__name__, (cls, LayeredDetector), {})
+            setattr(cls, "_original_cls", original_cls)
+        return super().__new__(cls)
+
+    def __init__(self, config: LayeredModelConfig = None, model: ModelBase = None, **kwargs):
+        super().__init__(config=self._resolve_args(config=config, model=model, **kwargs))
+
+    @classmethod
+    def _resolve_args(cls, config: LayeredModelConfig, model: ModelBase, **kwargs):
+        msg = f"Expected exactly one of `config.model` or `model` when creating LayeredModel {cls.__name__}."
+        if config is None and model is None:
+            raise RuntimeError(f"{msg} Received neither.")
+        elif config is not None and model is not None:
+            if config.model is None:
+                config.model = model
+            else:
+                raise RuntimeError(f"{msg} Received both.")
+        elif config is None:
+            config = cls.config_class(model=model, **kwargs)
+        return config
+
+    @property
+    def model(self):
+        return self.config.model
+
+    @model.setter
+    def model(self, model):
+        self.config.model = model
+
+    @property
+    def base_model(self):
+        return self.config.base_model
+
+    def reset(self):
+        self.model.reset()
+        self.__init__(config=self.config)
+
+    def __getstate__(self):
+        state = super().__getstate__()
+        state["model"] = self.model.__getstate__()
+        return state
+
+    def __setstate__(self, state):
+        if "model" in state:
+            model_state = state.pop("model")
+            self.model.__setstate__(model_state)
+        super().__setstate__(state)
+
+    def __reduce__(self):
+        state_dict = self.__getstate__()
+        config = state_dict.pop("config")
+        return getattr(self.__class__, "_original_cls", self.__class__), (config,), state_dict
+
+    def _save_state(self, state_dict: Dict[str, Any], filename: str = None, **save_config) -> Dict[str, Any]:
+        # don't save config for any of the sub-models
+        sub_state = state_dict
+        for d in range(self.config.depth):
+            sub_state.pop("config", None)
+            sub_state = sub_state["model"]
+        return super()._save_state(state_dict, filename, **save_config)
```

Comment:
The comment says that we don't save any of the sub-model configs, but it seems like this actually saves the base_model config?

Reply:
Good catch. I'll fix this.